skbio/metadata/_interval.py

```diff
@@ -0,0 +1,599 @@
+        self._interval_tree = IntervalTree()
+
+        # Indicates if the IntervalTree needs to be rebuilt.
+        self._is_stale_tree = False
+
+    def _reverse(self, length):
+        """Reverse complements IntervalMetadata object.
+
+        This operation reverses all of the interval coordinates.
+        For instance, this can be used to compare coordinates
+        in the forward strand to coordinates in the reversal strand.
+
+        Parameters
+        ----------
+        length : int
+            Largest end coordinate to perform reverse complement.
+            This typically corresponds to the length of sequence.
+        """
+        for f in self._intervals:
+            # staled doesn't need to be called, since the setter for
+            # Interval will take care of this
+            intvls = list(map(lambda x: (length-x[1], length-x[0]),
+                              reversed(f.locations)))
+            f.locations = intvls
+
+        # DONT' forget this!!!
+        self._is_stale_tree = True
+
+    def sort(self, ascending=True):
+        '''Sort intervals by their starting and ending coordinates.'''
+        self._intervals = sorted(
+            self._intervals,
+            key=lambda i: [i.locations[0][0], i.locations[-1][1]],
+            reverse=not ascending)
+
+    def add(self, locations, boundaries=None, metadata=None):
+        """Adds a feature to the metadata object.
+
+        This method creates a ``Interval`` object and insert it into
+        the ``IntervalMetadata`` object.
+
+        Parameters
+        ----------
+        locations : iterable of tuple of ints
+            A list of locations associated with the Interval object.
+        boundaries : iterable of tuple of bool
+            A list of boundaries associated with the locations.
+        metadata : dict
+            A dictionary of key word attributes associated with the
+            Interval object.
+
+        Returns
+        -------
+        Interval
+            The ``Interval`` object just added.
+
+        See Also
+        --------
+        skbio.metadata._add
+        """
+        # Add an interval to the tree. Note that the add functionality is
+        # built within the Interval constructor.
+        return Interval(interval_metadata=self,
+                        locations=locations,
+                        boundaries=boundaries,
+                        metadata=metadata)
+
+    def _rebuild_tree(self):
+        """Rebuilds the IntervalTree."""
+        self._interval_tree = IntervalTree()
+        for f in self._intervals:
+            for start, end in f.locations:
+                self._interval_tree.add(start, end, f)
+
+    def _query_interval(self, location):
+        """Fetches Interval objects that overlap with the location."""
+        _assert_valid_location(location)
+        # don't forget to update before query
+        if self._is_stale_tree:
+            self._rebuild_tree()
```

Comment:
should it be the responsibility of `_rebuild_tree` to set `_is_stale_tree` to `False`? The risk is a developer in the future may rebuild the tree and forget to unset the private member variable. 

One other way to possibly simplify and contain the logic further would be to always call `self._rebuild_tree()` here, and make it the responsibility of the `_rebuild_tree` method to determine if the tree should be build (i.e., check `_is_stale_tree`